[DO NOT MERGE] [stdlib] Random unification

Azoy opened the pull request:
This is the implementation for https://github.com/apple/swift-evolution/pull/760

Azoy committed:
Add shims for stdlib random

Azoy committed:
Initial random api

Azoy committed:
Use C syscall for I/O

1. Fixed an issue where integers would would result in an infinite loop if they were unsigned, or signed integers always returning negative numbers.
2. Fixed an issue with Bool initialization

Azoy committed:
Add shuffle functions

Azoy committed:
Add documentation to Random API

Azoy committed:
Fix a few typos within the documentation

Fixes more typos

Also states that the range for floating points is from 0 to 1 inclusive

Azoy committed:
Update API to reflect mailing list discussions

Remove unnecessary import

Make sure not to return upperBound on Range

Use SecRandomCopyBytes on older macOS

gparker42 commented:
Randomness distribution tests would be appropriate for the long tests.

Other basic tests are desirable: does the API work at all; if you ask for two large random numbers are they different; if you ask for a bunch of random numbers within a very small range do they all fall within that range.

gparker42 commented on stdlib/public/stubs/LibcShims.cpp:
```diff
@@ -213,3 +225,77 @@ swift::_swift_stdlib_cxx11_mt19937_uniform(__swift_uint32_t upper_bound) {
+void swift::_swift_stdlib_random(void *buf,
+                                __swift_ssize_t nbytes,
+                                __swift_uint32_t debug_flags) {
+  if (SecRandomCopyBytes(kSecRandomDefault, nbytes, buf) != 0) {
```
`SecRandomCopyBytes()` should be available on all of the Apple platforms supported by Swift. I don't think you need the `arc4random()` or `/dev/urandom` implementations there.

Azoy replied:
arc4random was suggested by Ben Cohen on the evolution mailing list here: https://lists.swift.org/pipermail/swift-evolution/Week-of-Mon-20171002/040195.html . I can substitute `SecRandomCopyBytes()` instead of reading `/dev/urandom`.

gparker42 replied:
My understanding is that `SecRandomCopyBytes()` is cryptographically superior to `arc4random()` on old Apple OS versions. I see no reason not to use it on all versions.

Azoy replied:
`arc4random(3)` was replaced in macOS 10.12 and iOS 10 to provide a more secure generator. I do indeed have the version checks to make sure to use `arc4random(3)` on those OS versions on and `SecRandomCopyBytes()` for older versions.

benrimmington replied:
* Should `arc4random_buf` vs `SecRandomCopyBytes` use [__builtin_available](https://clang.llvm.org/docs/LanguageExtensions.html#objective-c-available)?

* Should `__swift_ssize_t` be `__swift_size_t` instead?

* Can the `debug_flags` parameter be removed?
  Other APIs always call `fatalError` with zero flags.

```cpp
#if defined(__APPLE__)
#include <Security/Security.h>
SWIFT_RUNTIME_STDLIB_INTERFACE
void swift::_stdlib_random(void *buf, __swift_size_t nbytes) {
  if (__builtin_available(macOS 10.12, iOS 10, tvOS 10, watchOS 3, *)) {
    arc4random_buf(buf, nbytes);
  } else {
    OSStatus status {SecRandomCopyBytes(kSecRandomDefault, nbytes, buf)};
    if (status != errSecSuccess) {
      fatalError(0, "SecRandomCopyBytes failed with error %d\n", status);
    }
  }
}
#endif
```

Azoy committed:
Update API to match mailing list discussion, add tests

Added pick(_:) to collection
Added random(in:using:) to Randomizable
Added tests

Fix typo in Randomizable documentation

Rename pick to sampling

Move sampling below random

Azoy commented:
Added tests!

Azoy committed:
Update docs

Azoy committed:
Update from upstream

Azoy committed:
Use new Libc naming

Fix Random.swift with new Libc naming

Azoy committed:
Remove sampling

Azoy committed:
Update from upstream

Azoy committed:
gybify signed integer creation

Azoy committed:
Make FloatingPoint.random exclusive

Azoy committed:
Update from upstream

Azoy committed:
Refactor {Closed}Range.random

Fix FloatingPoint initialization

Precondition getting a random number from range

Azoy committed:
Merge branch 'master' into random-unification

Azoy committed:
Fix some doc typos

Azoy committed:
Make .random a function

Azoy committed:
Update API to reflect discussion

Make .random a function
Remove .random() in favor of .random(in:) for all numeric types

Azoy committed:
Merge branch 'master' of https://github.com/apple/swift into random-unification

benrimmington commented on stdlib/public/stubs/LibcShims.cpp:
```diff
@@ -329,3 +332,60 @@ swift::_stdlib_cxx11_mt19937_uniform(__swift_uint32_t upper_bound) {
+SWIFT_RUNTIME_STDLIB_INTERFACE
+void swift::_stdlib_random(void *buf,
+                                __swift_ssize_t nbytes,
+                                __swift_uint32_t debug_flags) {
```
* Should `arc4random_buf` vs `SecRandomCopyBytes` use [__builtin_available](https://clang.llvm.org/docs/LanguageExtensions.html#objective-c-available)?

* Should `__swift_ssize_t` be `__swift_size_t` instead?

* Can the `debug_flags` parameter be removed?
  Other APIs always call `fatalError` with zero flags.

* Reopening "/dev/urandom" for each `_stdlib_random` call seems inefficient.

```cpp
#if defined(__APPLE__)
#include <Security/Security.h>
SWIFT_RUNTIME_STDLIB_INTERFACE
void swift::_stdlib_random(void *buf, __swift_size_t nbytes) {
  if (__builtin_available(macOS 10.12, iOS 10, tvOS 10, watchOS 3, *)) {
    arc4random_buf(buf, nbytes);
  } else {
    OSStatus status {SecRandomCopyBytes(kSecRandomDefault, nbytes, buf)};
    if (status != errSecSuccess) {
      fatalError(0, "SecRandomCopyBytes failed with error %d\n", status);
    }
  }
}
#endif
```

Azoy replied:
Thanks!

Azoy committed:
Fix compile errors

Azoy committed:
Clean up _stdlib_random

Cleanup around API

Remove `.random()` requirement from `Collection`

Use generators

Optimize shuffle()

Thread safety for /dev/urandom

Remove {Closed}Range<BinaryFloatingPoint>.random()

Add Collection random requirement

Refactor _stdlib_random

Remove whitespace changes

Clean linux shim

benrimmington commented on stdlib/public/SwiftShims/LibcShims.h:
```diff
@@ -145,12 +145,16 @@ __swift_uint32_t _stdlib_cxx11_mt19937(void);
 
+// Random number for stdlib
+SWIFT_RUNTIME_STDLIB_INTERFACE
+void _stdlib_random(void *buf, __swift_size_t nbytes);
```
Should this function be `SWIFT_RUNTIME_STDLIB_INTERNAL` instead?
Should `@_inlineable` be removed from the `Random.next()` method?
<https://github.com/apple/swift/blob/master/stdlib/public/SwiftShims/Visibility.h>

Azoy replied:
Thanks!

benrimmington commented on stdlib/public/stubs/LibcShims.cpp:
```diff
@@ -329,3 +330,52 @@ swift::_stdlib_cxx11_mt19937_uniform(__swift_uint32_t upper_bound) {
+#include <sys/syscall.h>
+SWIFT_RUNTIME_STDLIB_INTERFACE
+void swift::_stdlib_random(void *buf, __swift_size_t nbytes) {
+  int result = syscall(SYS_getrandom, buf, nbytes, 0);
```
Why does `getrandom` need to be called via the `syscall`?
Should the `GRND_NONBLOCK` flag be used?
<http://www.man7.org/linux/man-pages/man2/getrandom.2.html>

Azoy replied:
I use `syscall` because `<sys/random.h>` doesn't exist on glibc versions 2.24 and earlier. I'd rather use `syscall` for all cases than doing another check for glibc on top of kernel version.

No, I think we want this to block until it can return a value.

benrimmington commented on stdlib/public/stubs/LibcShims.cpp:
```diff
@@ -329,3 +330,52 @@ swift::_stdlib_cxx11_mt19937_uniform(__swift_uint32_t upper_bound) {
+#else
+#include <mutex>
+static std::mutex m;
+static int fd;
```
The file descriptor could be moved inside the function.
C++11 has thread-safe lock-free initialization of local statics.
Or `SWIFT_LAZY_CONSTANT` uses `dispatch_once` and `std::call_once`.
<https://github.com/apple/swift/blob/master/include/swift/Basic/Lazy.h>

Azoy replied:
Thanks!

benrimmington commented on stdlib/public/stubs/LibcShims.cpp:
```diff
@@ -329,3 +330,52 @@ swift::_stdlib_cxx11_mt19937_uniform(__swift_uint32_t upper_bound) {
+      fatalError(0, "Fatal error: Unable to open /dev/urandom\n");
+    }
+  }
+  if (swift::_stdlib_read(fd, buf, nbytes) < 0) {
```
Reading from "/dev/urandom" (using `getrandom` or `_stdlib_read`)
can return fewer bytes than requested, so you might want to repeat
until the buffer is filled (or an error occurs).
<https://llvm.org/svn/llvm-project/libcxx/trunk/src/random.cpp>

Azoy replied:
Thanks!

natecook1000 commented on stdlib/public/core/CollectionAlgorithms.swift.gyb:
```diff
@@ -211,6 +211,56 @@ extension MutableCollection where Self : BidirectionalCollection {
+// shuffled()
+//===----------------------------------------------------------------------===//
+
+% for Self in ['Sequence', 'MutableCollection']:
```
We don't need a different version of `shuffled` for `MutableCollection`—just the `Sequence` version is fine. (I think you're following the example of `sorted` here, but that's actually redundant, too.)

natecook1000 reviewed (commented):
Other tests that might be good:

- Make sure we're getting all the values in a relatively short range of integers, for both `Range` and `ClosedRange`.
- Check that ranges and closed ranges work at the extrema of `Int64`, `UInt64`, `Int8`, and `UInt8`.
- Include an implementation of a seeded generator that can provide repeatable values, so we can test that the generator that's being passed as an argument is actually used for generation.

Azoy commented on stdlib/public/core/Collection.swift:
```diff
@@ -62,7 +62,7 @@ public typealias Indexable = Collection
 ///             }
 ///         }
-///         
+///
```
Done!

natecook1000 replied:
Nice work, thanks!

benrimmington commented on stdlib/public/stubs/LibcShims.cpp:
```diff
@@ -329,3 +330,52 @@ swift::_stdlib_cxx11_mt19937_uniform(__swift_uint32_t upper_bound) {
+  }
+}
+#endif
+#endif
```
The `buf` and `nbytes` need to be updated within the loop.

```cpp
#if defined(__linux__)
#include <linux/version.h>
#include <sys/syscall.h>
SWIFT_RUNTIME_STDLIB_INTERNAL
void swift::_stdlib_random(void *buf, __swift_size_t nbytes) {
#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,17,0)
  auto _read = [&]() -> __swift_ssize_t {
    return syscall(SYS_getrandom, buf, nbytes, 0);
  };
#else // LINUX_VERSION_CODE
  auto _read = [&]() -> __swift_ssize_t {
    static const int fd = _stdlib_open("/dev/urandom", O_RDONLY);
    if (fd < 0) {
      fatalError(0, "Unable to open '/dev/urandom'\n");
    }
    return _stdlib_read(fd, buf, nbytes);
  };
#endif // LINUX_VERSION_CODE
  while (nbytes > 0) {
    auto result = _read();
    if (result < 1) {
      if (errno == EINTR) { continue; }
      fatalError(0, "Unable to read '/dev/urandom'\n");
    }
    buf = static_cast<uint8_t *>(buf) + result;
    nbytes -= result;
  }
}
#endif
```

Is reading from a shared file descriptor thread-safe?

Azoy replied:
Sorry, my C++ isn't the greatest. Yes, reading from a shared file descriptor is thread-safe.

benrimmington commented:
There are some typos (`bytes` vs `nbytes`) in the `_stdlib_random` function.
The internal C++ code might be merged sooner as a separate pull request?
It could be used in [GlobalObjects.cpp](https://github.com/apple/swift/blob/99229582ee3d4ed015e076e67e9b050ee5ed5441/stdlib/public/stubs/GlobalObjects.cpp#L109-L120) to replace `std::random_device`, etc.

Azoy committed:
Merge branch 'master' of https://github.com/apple/swift into random-unification

- Added more tests (more to come)

Azoy committed:
Revert "Merge branch 'master' of https://github.com/apple/swift into random-unification"

This reverts commit b0ca3d4639d1c678a7330bb2db3d9c41dbf2e523.

Azoy committed:
Fix CollectionAlgorithm conflict

Azoy committed:
Add shuffle and more tests

huonw commented on stdlib/public/core/Bool.swift:
```diff
@@ -87,6 +87,19 @@ public struct Bool {
+  public static func random(
+    using generator: RandomNumberGenerator = Random.default
+  ) -> Bool {
+    let random = Int8.random(in: 0 ... 1, using: generator)
```
Could you say why you're using this relatively complicated implementation, vs. something like `generator.next() % 2 == 0`?

huonw commented on test/stdlib/Random.swift:
```diff
@@ -0,0 +1,153 @@
+
+func chi2Test(_ samples: [Double]) -> Bool {
+  let upperBound = 50
+  let numberOfTrials = 500_000
```
These two could probably be either a parameter or a global, since they're shared (and need to be the same as) the values in the test below.

In fact, they're both somewhat redundant:  `upperBound == samples.count` and `numberOfTrails = samples.reduce(0, +)`. And, since the DoF is fixed (by the hardcoded confidence interval), it might be worth having it as an assert instead:

```swift
func chi2Test(_ samples: [Double]) -> Bool {
    precondition(samples.count == 50, "confidence interval requires 50 samples")
    let expected = samples.reduce(0, +) / Double(samples.count)
    // ...
}
```

huonw commented on test/stdlib/Random.swift:
```diff
@@ -0,0 +1,84 @@
+  let numberOfTrials = 500_000
+  let expected = Double(numberOfTrials / upperBound)
+  let cvLow = 28.9 // 1% with a degree of freedom of (50 - 1)
+  let cvHigh = 74.9 // 99% with a degree of freedom of (50 - 1)
```
And, we hammer tests: by my count on ci.swift.org, there have been at least 30 smoke test runs in the last 24 hours, each of which runs on both macOS and Linux, meaning there would be a 70% chance of at least one of them spuriously failing due to this test. (And, this isn't including other testing like non-smoke tests and all the test runs developers run on their own machines.)

Azoy replied:
So would an appropriate percentage be 0.1%?

Azoy committed:
Provide finishing tests and suggestions

natecook1000 commented:
Good catch! Everything should be expressed in `Range` or `ClosedRange` if it isn't already.

Azoy committed:
Merge branch 'master' of https://github.com/apple/swift into random-unification

Azoy committed:
Remove refs to Countable ranges

Revert to checking if T is > UInt64

benrimmington committed:
Add `_stdlib_random` for more platforms (#1)

* Remove refs to Countable ranges

* Add `_stdlib_random` for more platforms

* Use `getrandom` (if available) for Android, Cygwin

* Reorder the `_stdlib_random` functions

* Also include <features.h> on Linux

* Add `#error TODO` in `_stdlib_random` for Windows

* Colon after Fatal Error

Performance improvement for Random

gybify ranges

Fix typo in 'basic random numbers'
Add _stdlib_random as a testable method

Switch to generic constraints

Hopefully link against bcrypt

benrimmington commented on test/stdlib/Random.swift:
```diff
@@ -0,0 +1,225 @@
+  let randomDouble1 = Double.random(in: 0 ..< 1)
+  expectTrue(randomDouble1 < 1 && randomDouble1 >= 0)
+  let randomDouble2 = Double.random(in: 0 ..< 1)
+  expectTrue(randomDouble1 < 1 && randomDouble2 >= 0)
```
Line 18 should be `expectTrue(randomDouble2 < 1 && randomDouble2 >= 0)`

Maybe the "basic random numbers" test should use two arrays of numbers?

benrimmington commented:
For the `ClosedRange.random(using:)` and `Range.random(using:)` methods:

* Can the latter call the former (to reduce code duplication)?

* Can they be moved to a GYB file (to reduce doc-comment duplication)?

* Can they be removed, by using `RangeExpression` in the `FixedWidthInteger.random(in:using:)` method?

Azoy commented on stdlib/public/core/CMakeLists.txt:
```diff
@@ -220,6 +222,10 @@ if(CMAKE_SYSTEM_NAME STREQUAL "FreeBSD")
+if(CMAKE_SYSTEM_NAME STREQUAL "Windows")
+  list(APPEND swift_core_link_flags "$ENV{SystemRoot}/system32/bcrypt.dll")
+endif()
+
```
@benrimmington Does this look correct?

benrimmington replied:
@Azoy, I think the `Bcrypt.lib` import library is needed.

<https://msdn.microsoft.com/en-us/library/windows/desktop/ms681914(v=vs.85).aspx>

Are you able to test on Windows? If not, shall we ask @compnerd for help?

benrimmington replied:
@compnerd, please see Azoy/swift@b2ef08f51de37a18d61d2c3869d116d0dc62dfe2 in Azoy/swift#2.

It uses `#pragma comment(lib, "Bcrypt.lib")` to add the import library.

It also uses `BCRYPT_SUCCESS` instead of `NT_SUCCESS`, so the extra defines/includes (`WIN32_NO_STATUS`, `<Ntstatus.h>`, `<Ntdef.h>`) aren't needed AFAIK.

orent commented:
On Linux, libc versions and kernel versions are not tightly coupled and can be skewed significantly in either direction. 

Code compiled with a new libc should be able to run on older kernels where get-random() is not available rather than failing. For example, a stable system with an older kernel may run chroots or containers that have the latest and greatest usermode system. 

Code may be intentionally linked on old libc for portability but expected to execute on new kernel. Why not let it enjoy the getrandom syscall?

I suggest implementing this selection at runtime with fallback, not as compile-time #ifdefs.

benrimmington commented:
> Code compiled with a new libc should be able to run on older kernels where get-random() is not available rather than failing.

@orent, it's not supposed to fail. It should use "/dev/urandom":

* if `getrandom` isn't available (Android, Cygwin, Linux);
* or as a last resort (FreeBSD, Haiku, etc).

> Code may be intentionally linked on old libc for portability but expected to execute on new kernel. Why not let it enjoy the getrandom syscall?

@Azoy originally used `syscall(SYS_getrandom, ...)` but I changed it to a library call, so that Cygwin can use the same implementation.

> I suggest implementing this selection at runtime with fallback, not as compile-time #ifdefs.

How, exactly? Ideally, there shouldn't be almost-duplicate versions of `_stdlib_random` for Android/Linux and Cygwin.

orent commented:
> We require that the kernel version be >= 3.17 as this was the release that introduced the getrandom(2) system call. We also require that glibc be >= 2.25 because this release exposed the <sys/random.h> header.

I see no good reason for Swift programs to require a specific minimal kernel version. Leave such considerations to the underlying libc as much as possible. It is ok to build the latest glibc with old kernel headers (in that case glibc will provide a getrandom usermode stub that always return ENOSYS). It is ok to build the latest glibc with new kernel headers but run it on a very old kernel (as far back as the 2.4 line). 

When running on an old kernel, some glibc syscalls will fail with ENOSYS and the caller of that glibc function should fail gracefully or provide a fallback to an alternative implementation.

As currently implemented, a Swift program compiled with glibc >=2.25 will always try to use getrandom() and fail when executed on kernel <3.17, even when /dev/urandom is available.

Also, certain embedded platforms may have a new kernel and a new glibc, but for some reason use a cross-compilation toolchain with a copy glibc that was compiled with older kernel headers. In that case, Swift will produce programs that are completely unuseable on any kernel version. 

The fallback from getrandom(errno == ENOSYS) to /dev/urandom must be implemented at runtime.

benrimmington commented:
@orent:

> It is ok to build the latest glibc with new kernel headers but run it on a very old kernel (as far back as the 2.4 line).
>
> When running on an old kernel, some glibc syscalls will fail with ENOSYS and the caller of that glibc function should fail gracefully or provide a fallback to an alternative implementation.

The [`getrandom` in glibc](https://sourceware.org/git/?p=glibc.git;a=blob;f=sysdeps/unix/sysv/linux/getrandom.c;hb=HEAD) uses `#ifdef __NR_getrandom`, so I don't think your `errno == ENOSYS` suggestion will make any difference. The selection of system call or usermode stub has already happened at compile time.

> As currently implemented, a Swift program compiled with glibc >=2.25 will always try to use getrandom() and fail when executed on kernel <3.17, even when /dev/urandom is available.

The current implementation in this pull request uses:

```c
#if defined(__linux__)
#  include <linux/version.h>
#  if (LINUX_VERSION_CODE >= KERNEL_VERSION(3,17,0))
#    include <features.h>
#    if defined(__BIONIC__) || (defined(__GLIBC_PREREQ) && __GLIBC_PREREQ(2,25))
#      include <sys/random.h>
#      define SWIFT_STDLIB_USING_GETRANDOM
#    endif
#  endif
#endif
```

In other words, `getrandom` will only be used if compiled with Linux 3.17 *and* glibc 2.25 (or later). Otherwise, "/dev/urandom" will be used instead.

Do you want to compile Swift with the latest kernel headers, and then deploy back to an older kernel? Is this supported? Why does the [download](https://swift.org/download/) page have three separate Ubuntu releases, and not a single release for all supported versions?

Note that libc++ uses "/dev/urandom" (see `_LIBCPP_USING_DEV_RANDOM`) by default, and doesn't use `getrandom` at all:

* <https://llvm.org/svn/llvm-project/libcxx/trunk/include/__config>
* <https://llvm.org/svn/llvm-project/libcxx/trunk/include/random>
* <https://llvm.org/svn/llvm-project/libcxx/trunk/src/random.cpp>

Azoy committed:
Merge branch 'master' of https://github.com/apple/swift into random-unification

otex commented:
The simplest and most reliable way to find if your libc has getrandom() is probably #ifdef GRND_RANDOM. This macro has no other purpose other than being passed as a flag to getrandom() . There is no need for any assumptions about GLIBC, BIONIC, or specific version numbers.

But even if your libc has getrandom() you should not assume that the kernel necessarily supports the syscall. LINUX_VERSION_CODE is the version of kernel headers that glibc was built with and the highest it is aware of. But executables linked with this glibc will happily on newer linux kernels, older linux kernels, kernels with custom configurations that have certain features disabled and even things that are not a linux kernel like the linux syscall API emulation layers of Windows, BSD or Solaris. All these targets may or may not support specific flags and syscalls. You should expect errnos like EUNSUPP or ENOSYS to be returned at runtime on some environments.

I understand that your reference linux platforms are a number of specific Ubuntu versions. In this case you can expect getrandom to be either supported on both glibc and the kernel or not at all. But it is not a good idea to encode such assumptions about versions into the code. Swift code will run in environments where glibc supports getrandom but the kernel does not. You should provide a fallback to /dev/urandom at runtime, not with #ifdefs.

orent commented:
Even better: use arc4random on all platforms. On modern OSX it’s available in libc. On Linux you can find it in the libbsd package that Swift already depends on. On other platforms you can include a vendored copy of libbsd. 

It is the right thing to do because arc4random is a fast userspace pseudorandom generator that is seeded from a source of high-quality entropy. Using the kernel’s randomness API for everything is slow. 

And finally, libbsd already handles all this platform-specific stuff for you. Instead of the getrandom syscall it can use an obscure sysctl kernel API for generating uuids. It comes from the same randomness pool, works even when /dev/urandom is unavailable and is supported on much older kernels.

benrimmington commented:
> The simplest and most reliable way to find if your libc has getrandom() is probably #ifdef GRND_RANDOM.

Good idea! We can also use one of the [*include file checking macros*](https://clang.llvm.org/docs/LanguageExtensions.html#include-file-checking-macros):

```c
#if __has_include(<sys/random.h>)
#  include <sys/random.h>
#endif
```

> Even better: use arc4random on all platforms.

This was rejected, see the comments in <https://github.com/Azoy/swift/pull/1>.

orent commented:
Recent versions of arc4random use chacha, not rc4. Faster and secure. Vendor it on all platforms, with the exception of known-good Apple platforms.

There should be no reason for a process to ever read more than 32 bytes from the OS entropy source. Use it once for seeding arc4random and you are done. A syscall for every random number is not an acceptable overhead.

Azoy commented:
@benrimmington it seems sensible to me to at least fallback on reading `/dev/urandom` when we receive `ENOSYS` or `EUNSUPP` even when we don't expect it. We probably need to take another look at Linux's implementation (we should probably use a mutex when reading `/dev/urandom` directly).

@orent I agree `arc4random` in libbsd uses chacha, however platforms such as Ubuntu 14.04 still rely on older versions of libbsd which do in fact use the rc4 version of `arc4random`. Swift does use libbsd as a dependency in terms of development, but if you're just downloading the Swift binaries from `swift.org` then libbsd is not a requirement. I think it would be a huge undertaking adding another dependency for Linux systems.

Azoy committed:
Merge branch 'master' of https://github.com/apple/swift into random-unification

orent commented:
So vendor it. Statically. On all platforms, including all Linux versions, except known-good Apple platforms.

You need the functionality of arc4random - wrapping a slow, high quality OS entropy source with a fast userspace random generator. And for the same price you also get the all he platform-specific mess solved for you by someone else for free.

Azoy commented:
I don't have any authority over whether or not we should vendor it. What we can do now is the current design which we have authority over and is somewhat feasible for the time being. You can ask @tkremenek  (hate to bother you) as he would be the one to have the final say is something like that.

@natecook1000 I'm getting some really weird compile time errors after pulling from master. Specifically this line [here](https://github.com/apple/swift/pull/12772/files#diff-d1f4823ad2f0838ef09162841df28256R2431). `error: argument passed to call that takes no arguments`. I didn't receive this previously, and the function that it's supposed to call is defined right above it. Any ideas?

natecook1000 commented:
> I'm getting some really weird compile time errors after pulling from master. Specifically this line here. error: argument passed to call that takes no arguments. I didn't receive this previously, and the function that it's supposed to call is defined right above it. Any ideas?

That is odd — my normal steps when running into things like that are to run `utils/update-checkout` to make sure the related dependencies are all up to date, and then do a clean build.

Azoy commented:
Yeah, so after `utils/update-checkout` and a few clean builds, it still fails to compile. I've been able to mitigate this issue by changing the `Range.random(using:)` (which is defined right above the `Range.random()`) to the following:
```swift
public func random<T: RandomNumberGenerator>(
  using generator: T
) -> Element?
// to
public func random(
  using generator: RandomNumberGenerator
) -> Element?
```
This is interesting because none of the other APIs that follow this design have this issue. Maybe I'm not seeing something?

Azoy committed:
Merge branch 'master' of https://github.com/apple/swift into random-unification

orent commented:
You already have a version of arc4random with multiplatform getentropy. In-house, all nice and vetted by your lawyers under an Apple license. I guess this should make the vendoring issue easier.

https://opensource.apple.com/source/network_cmds/network_cmds-543/APPLE_LICENSE
https://opensource.apple.com/source/network_cmds/network_cmds-543/unbound/compat/arc4random.c
https://opensource.apple.com/source/network_cmds/network_cmds-543/unbound/compat/getentropy_linux.c

NevinBR reviewed (changes requested):
next(upperBound:) distribution is not uniform.

Azoy committed:
Fix some implementation details

1. Uniform distribution is now uniform
2. Apply Jens' method for uniform floats

Fix a lineable attribute

Azoy commented:
@NevinBR it is now!

natecook1000 committed:
Revise documentation, add benchmarks (#3)

* [stdlib] Revise documentation for new random APIs

* [stdlib] Fix constraints on random integer generation

* [test] Isolate failing Random test

* [benchmark] Add benchmarks for new random APIs

Fix Float80 test

orent commented:
The libbsd on some platforms is not up to date. Their arc4random function may actually use rc4 (now considered weak) and the entropy seeding may not be up to the latest standards. That is why I suggested using the OS implementation on known-good platforms and a bundled version on others.

This is similar to the approach taken by the OpenBSD project. Their OpenSSH and other tools are now critical global infrastructure components. Instead of having a tangle of #ifdefs they implement to a single reference platform (OpenBSD) and provide platform-specific libraries to emulate its functions on other environments.

benrimmington commented:
But is libbsd still a dependency for the standard library? apple/swift#1965

Azoy committed:
Value type generators

random -> randomElement

Fix some docs

One more doc fix

Doc fixes & bool fix

Use computed over explicit

natecook1000 commented:
@swift-ci Please smoke test

natecook1000 commented:
@swift-ci Please smoke benchmark

natecook1000 commented on stdlib/public/core/Random.swift:
```diff
@@ -0,0 +1,152 @@
+/// cryptographically secure algorithm whenever possible.
+public struct Random : RandomNumberGenerator {
+  /// The shared, default instance of the `Range` random number generator.
+  public static var `default` = Random()
```
@jckarter Does this need to be the implementation you gave here? https://forums.swift.org/t/se-0202-random-unification/11313/171

jckarter replied:
The core team decided that passing an explicit default is likely rare enough not to need the no-op setter to begin with.

Azoy commented:
@natecook1000 I'm not understanding why `Range` is in this description: `/// The shared, default instance of the Range random number generator.` Also, should I update the documentation to include the initial implementation of `Random` for different platforms? (`The standard library should document the implementation used on each platform.`)

natecook1000 commented:
That… looks like a typo—should be `Random`. And yes, it would be great to get that initial information in there.

benrimmington committed:
Consolidate `_stdlib_random` functions (#2)

* Use the `__has_include` and `GRND_RANDOM` macros

* Use `getentropy` instead of `getrandom`

* Use `std::min` from the <algorithm> header

* Move `#if` out of the `_stdlib_random` function

* Use `getrandom` with "/dev/urandom" fallback

* Use `#pragma comment` to import "Bcrypt.lib"

* <https://docs.microsoft.com/en-us/cpp/preprocessor/comment-c-cpp>
* <https://clang.llvm.org/docs/UsersManual.html#microsoft-extensions>

* Use "/dev/urandom" instead of `SecRandomCopyBytes`

* Use `swift::StaticMutex` for shared "/dev/urandom"

* Add `getrandom_available`; use `O_CLOEXEC` flag

Add platform impl docs

Update copyrights

Fix docs

Add _stdlib_random test

Update _stdlib_random test

Add missing &

Notice about _stdlib_random

Fix docs

Guard on upperBound = 0

Test full range of 8 bit integers

Remove some gyb

benrimmington commented on stdlib/public/core/Random.swift:
```diff
@@ -118,8 +118,19 @@ extension RandomNumberGenerator {
+///
+/// - `Darwin` devices use `arc4random_buf(3)` for macOS 10.12 or higher, iOS 10
+///   or higher, tvOS 10 or higher, and watchOS 3 or higher. For older versions
+///   of these operating systems, they read from `/dev/urandom`.
```
Could this be "Apple platforms" instead of "Darwin devices"? For example:

> Apple platforms use `arc4random_buf` (with the NIST-approved AES cipher, but not the RC4 algorithm) or `/dev/urandom`.

Azoy replied:
How does this sound, "Apple platforms use `arc4random_buf(3)` for newer versions of their OS, and for older versions they will read from `/dev/urandom`."?

benrimmington replied:
Sounds good. Do you need to mention any others ([FreeBSD, Haiku, PS4](https://github.com/apple/swift/blob/master/lib/Basic/LangOptions.cpp)) or can they be included with the `getrandom` description?

Azoy commented on stdlib/public/core/Random.swift:
```diff
@@ -0,0 +1,166 @@
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
```
I'm assuming any files that were updated in this pr need to also update their copyright year?

benrimmington replied:
I suppose so, although I haven't noticed any other pull requests which do that.

benrimmington reviewed (commented):
Here's a quick review of the documentation.

natecook1000 commented on stdlib/public/core/CollectionAlgorithms.swift:
```diff
@@ -260,6 +260,115 @@ extension MutableCollection where Self : BidirectionalCollection {
+  /// - Parameter generator: The random number generator to use when shuffling
+  ///   the collection.
+  ///
+  /// - Complexity: O(*n*)
```
Hmmm. If the collection isn't random-access, this ends up being O(n^2), since `index(_:offsetBy:)` gets called *n* times. For random-access collections, it's O(n), since there are `count - 1` swaps.

benrimmington replied:
Accessing the `count` property can be O(*n*), so could you store it locally with name shadowing?

```swift
let count = self.count
guard count > 1 else { return }
// etc.
```

natecook1000 replied:
Yes, that's an optimization that should be added at some point.

Azoy replied:
or should I do something like this:
```swift
/// - Complexity: O(*n*) if the collection conforms to
///   `RandomAccessCollection`; otherwise, O(*n^2*), where *n* is the length
///   of the collection.
```

natecook1000 replied:
To be honest, I was thinking we should probably give it the same constraints as `sort`, which is to add `RandomAccessCollection` as a requirement for the in-place shuffling methods.

@airspeedswift What do you think? Does parity with `sort` make sense?

natecook1000 commented on stdlib/public/core/CollectionAlgorithms.swift:
```diff
@@ -260,6 +260,115 @@ extension MutableCollection where Self : BidirectionalCollection {
+  ///   the sequence.
+  /// - Returns: An array of this sequence's elements in a shuffled order.
+  ///
+  /// - Complexity: O(*n*)
```
The nonmutating versions will always be O(n), since it copies the sequence into an array first, which is random-access.

benrimmington replied:
Why does it copy the sequence into a `ContiguousArray`, shuffle that, and then copy the result into an `Array`?

natecook1000 replied:
The copy into `ContiguousArray` forces an eager bridge if the sequence is an `NSArray`. The copy into `Array` at the end reuses the same buffer, so it kind of just acts like a cast.

benrimmington commented on stdlib/public/core/FloatingPoint.swift.gyb:
```diff
@@ -2374,6 +2374,113 @@ extension BinaryFloatingPoint {
+  ///     // Prints "14.2286325689993"
+  ///     // Prints "13.1485686260762"
+  ///
+  /// The `random(using:)` static method chooses a random value from a
```
`random(using:)` => `random(in:using:)`

benrimmington commented:
Should this pull request be `@inlinable` audited, or will that happen later for Swift 5?

benrimmington commented on test/stdlib/Random.swift:
```diff
@@ -5,6 +5,31 @@ import StdlibUnittest
+  }
+  
+  // buffer1 == buffer2
+  expectTrue(buffer1.elementsEqual(buffer2))
```
```swift
for count in [100, 1000] {
    // Initialize two large buffers.
    var bytes1 = [UInt8](repeating: 0, count: count)
    var bytes2 = [UInt8](repeating: 0, count: count)
    expectEqual(bytes1, bytes2)

    // Fill with random bytes.
    bytes1.withUnsafeMutableBytes({ _stdlib_random($0) })
    bytes2.withUnsafeMutableBytes({ _stdlib_random($0) })
    expectNotEqual(bytes1, bytes2)

    // Test for unfilled repeating zeros.
    var zeros = [UInt8](repeating: 0, count: 10)
    expectNil(memmem(&bytes1, bytes1.count, &zeros, zeros.count))
    expectNil(memmem(&bytes2, bytes2.count, &zeros, zeros.count))
}
```

I've used `memmem` in this example, but is there a stdlib API to find a subsequence?

benrimmington replied:
Here's a unit test which also shows how `_stdlib_random` might work in the Foundation overlay.

```swift
import Foundation

extension Data {
  static func _random(count: Int) -> Data {
    var result = Data(count: count)
    result.withUnsafeMutableBytes {
      (start: UnsafeMutablePointer<UInt8>) -> Void in
      let start = UnsafeMutableRawPointer(start)
      let bytes = UnsafeMutableRawBufferPointer(start: start, count: count)
      _stdlib_random(bytes)
    }
    return result
  }
}

let zeros = Data(count: 10)
for count in [100, 1000] {
  let data1 = Data._random(count: count)
  let data2 = Data._random(count: count)
  expectNil(data1.range(of: zeros))
  expectNil(data2.range(of: zeros))
  expectNotEqual(data1, data2)
}
```

Should it actually be `_swift_stdlib_random` with the `SWIFT_RUNTIME_STDLIB_SPI` attribute?

Azoy replied:
We can keep it `_stdlib_random` with `SWIFT_RUNTIME_STDLIB_INTERNAL` atm because I plan to follow up with another proposal to fill buffers. The solution atm isn’t optimal (calling `Random.default.next()` many times), but it lets us keep stdlib visibility. In your memmem example, couldn’t you call `elementsEqual(_:)` because you’re checking if all elements are equal in sequential order?

benrimmington replied:
* [`Sequence.elementsEqual`](https://github.com/apple/swift/blob/8ef98b8dd2dd4519c2a94f491c68f1254daa4ac3/stdlib/public/core/SequenceAlgorithms.swift#L273-L345) will iterate both sequences, even if they have a different number of elements.

* [`Array.==`](https://github.com/apple/swift/blob/8ef98b8dd2dd4519c2a94f491c68f1254daa4ac3/stdlib/public/core/Arrays.swift.gyb#L2198-L2263) returns early if the `count` is different, or if the buffer identity is the same.

* [`Data.==`](https://github.com/apple/swift/blob/f38192d7b7afe08a2e26cf8077769183cd18d063/stdlib/public/SDK/Foundation/Data.swift#L1844-L1870) also returns early, and uses the `memcmp` API (which might be faster).

In this example, the arrays have the same number of elements in different buffers, so `elementsEqual` and `==` will do the same thing.

The `memmem` API might not be available on all platforms.

benrimmington commented on stdlib/public/core/Collection.swift:
```diff
@@ -1003,6 +1022,54 @@ extension Collection {
+    let random = generator.next(upperBound: UInt(count))
+    let index = self.index(
+      startIndex,
+      offsetBy: numericCast(random)
```
However, this would be indirectly calling `Range.randomElement(using:)` in the default implementation of `Collection.randomElement(using:)`, which is a protocol requirement.

benrimmington replied:
Indirectly calling `Range.randomElement(using:)` inside `Collection.randomElement(using:)` won't be recursive. But will it be slower than `next(upperBound:)` and `numericCast(_:)`?

Azoy replied:
My initial reaction is yes it will be slower, but it should be minuscule however.

benrimmington commented on stdlib/public/stubs/LibcShims.cpp:
```diff
@@ -308,3 +318,67 @@ swift::_stdlib_cxx11_mt19937_uniform(__swift_uint32_t upper_bound) {
   std::uniform_int_distribution<__swift_uint32_t> RandomUniform(0, upper_bound);
   return RandomUniform(getGlobalMT19937());
 }
+
```
You could add a reminder here to update the default `RandomNumberGenerator` documentation (if `_stdlib_random` is changed).

Azoy commented on stdlib/public/core/CollectionAlgorithms.swift:
```diff
@@ -260,6 +260,114 @@ extension MutableCollection where Self : BidirectionalCollection {
+    var currentIndex = startIndex
+    while amount > 1 {
+      let random = generator.next(upperBound: UInt(amount))
+      amount -= 1
```
Is there any benefit to using an alternative here rather than the first one that you pointed out (which is what is currently implemented)?

benrimmington replied:
The only benefit is less lines of code, and possibly the intention is more obvious.

Azoy commented:
@benrimmington @natecook1000 are there anymore implementation details that we need to work out (Besides hearing a response from @airspeedswift regarding constraining the shuffle implementation to `RandomAccessCollection` to have parity with `sort`)?

benrimmington commented:
I've already asked about adding the generic `next()` and `next(upperBound:)` as protocol requirements.

But the problem is those APIs aren't tested with types larger than `UInt64`, and even using `DoubleWidth` won't test the condition where `T.bitWidth.quotientAndRemainder(dividingBy: UInt64.bitWidth)` returns a non-zero remainder.

I also wonder if `next(upperBound:)` could be private API, because it seems unlikely that a generator will be able to implement it differently. For example, the `arc4random_uniform` API can't be used, because `next(upperBound:)` has a generic return type.

I haven't reviewed `BinaryFloatingPoint.random(in:using:)` or `Range.randomElement(using:)` because I don't understand them.

benrimmington commented on stdlib/public/core/Random.swift:
```diff
@@ -0,0 +1,166 @@
+  @inlinable
+  public mutating func next<T: FixedWidthInteger & UnsignedInteger>(
+    upperBound: T
+  ) -> T {
```
Does this need one of the following?

* `guard upperBound != 0 else { return 0 }`
* `precondition(upperBound != 0,
  "Can't get random value with lowerBound == upperBound")`

Azoy replied:
Hmm, I think the first one would make the most sense rather than trapping. Thanks!

Azoy commented:
We can't move `next()` or `next(upperBound:)` to requirements yet. I've asked Ben and he said it would be best to follow up with another proposal (including something like `fillBuffer(_:)` which I'm writing up now). 

As for testing `next()`, to me it seems unreasonable to even expect someone to create something like `UInt83`, maybe I'm wrong 🤔 . I think once `DoubleWidth` gets released we should add new tests make sure those get filled up.

benrimmington commented:
* [`UInt0`][UInt0] and [`UInt72`][UInt72] have been mentioned on the forums.
* [`Builtin.Int1`][Int1], [`Builtin.Int63`][Int63], and [`Builtin.Int80`][Int80] are used internally.

[UInt0]:  <https://forums.swift.org/t/how-do-i-make-a-uint0/9516>
[UInt72]: <https://forums.swift.org/t/can-anyone-quick-check-this-arbitrary-precision-code/9925>
[Int1]:   <https://github.com/apple/swift/blob/37532071e4df997a0214f37651063894b6e9508d/stdlib/public/core/Bool.swift#L65>
[Int63]:  <https://github.com/apple/swift/blob/37532071e4df997a0214f37651063894b6e9508d/stdlib/public/core/Character.swift#L74>
[Int80]:  <https://github.com/apple/swift/blob/37532071e4df997a0214f37651063894b6e9508d/test/stdlib/FloatingPoint.swift.gyb#L32-L50>

benrimmington commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -2483,6 +2483,172 @@ ${assignmentOperatorComment(x.operator, False)}
+  @inlinable
+  public func randomElement<T: RandomNumberGenerator>(
+    using generator: inout T
+  ) -> Element? {
```
I think the following might be possible, to avoid the GYB code.

```swift
extension ${Range} where
  Bound: FixedWidthInteger,
  Bound.Stride: SignedInteger,
  Bound.Magnitude: UnsignedInteger
{

  internal var _magnitude: Bound.Magnitude? {
    guard !(contains(.min) && contains(.max)) else {
      return nil
    }
    let magnitude: Bound.Magnitude
    switch (lowerBound.signum(), upperBound.signum()) {
    case (-1, -1):
      magnitude = lowerBound.magnitude - upperBound.magnitude
    case (1, 1):
      magnitude = upperBound.magnitude - lowerBound.magnitude
    default:
      magnitude = lowerBound.magnitude + upperBound.magnitude
    }
    return contains(upperBound) ? magnitude + 1 : magnitude
  }

  public func randomElement<T: RandomNumberGenerator>(
    using generator: inout T
  ) -> Element? {
    guard !isEmpty else {
      return nil
    }
    let randomDistance: Bound.Magnitude
    if let maxDistance = _magnitude {
      randomDistance = generator.next(upperBound: maxDistance)
    } else {
      randomDistance = generator.next()
    }
    return lowerBound &+ Bound(truncatingIfNeeded: randomDistance)
  }
}
```

The overflow addition is a bit obscure, but it seems to work.

Is there a `Bound.bitWidth == Bound.Magnitude.bitWidth` guarantee?

natecook1000 replied:
I don't think we need to treat avoiding GYB in a GYB'd file as a goal, particularly if it ends up adding runtime branches. If it makes sense to extract some of this to a common routine, that's great, but GYB should generate the code you would write if you were writing out each version specifically. (Here and below.)

benrimmington commented on stdlib/public/core/FloatingPoint.swift.gyb:
```diff
@@ -2374,6 +2374,113 @@ extension BinaryFloatingPoint {
+      range.lowerBound != range.upperBound,
+      "Can't get random value with lowerBound == upperBound"
+    )
+%   end
```
Instead of comparing bounds inside a GYB block, you might be able to use:

```swift
_precondition(!range.isEmpty, "Can't get random value with empty range")
```

With closed ranges, the [`isEmpty`](https://developer.apple.com/documentation/swift/closedrange/1779189-isempty) property is always false.

benrimmington commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -2483,6 +2483,172 @@ ${assignmentOperatorComment(x.operator, False)}
+      range.lowerBound != range.upperBound,
+      "Can't get random value with lowerBound == upperBound"
+    )
+%   end
```
Ditto:

```swift
_precondition(!range.isEmpty, "Can't get random value with empty range")
```

benrimmington commented on test/stdlib/Random.swift:
```diff
@@ -0,0 +1,246 @@
+
+// Basic random numbers
+
+RandomTests.test("basic random numbers") {
```
Could this also test the full ranges (min to max) of `Int8` and `UInt8`?

The number of iterations would need to increase (to 100_000 or 1_000_000 perhaps).

Azoy replied:
I'm not sure I'm following when you say "test the full ranges".

benrimmington replied:
Sorry, I attached the previous comment to the wrong test. I meant the "random integers in ranges" test, using the `integerRangeTest` helper function. It's currently testing for the min/max ± 10, but it might be good to also test for the entire 256 elements of the 8-bit closed ranges (and maybe half-open ranges).

* `(Int8.min ... .max)` and `(UInt8.min ... .max)`
* `(Int8.min ..< .max)` and `(UInt8.min ..< .max)`

Azoy replied:
I've done a few tests with 10_000 iterations and it hasn't failed once, so I'm going to set it to that initially. Let me know if we still need to increase it just in case.

benrimmington replied:
You might be able to simplify `integerRangeTest` by adding a couple more helper functions (one taking a half-open range, another taking a closed range). The number of iterations could then be `for _ in 0 ..< (100 * range.count)` for example. The body of the existing `integerRangeTest` would be:

```swift
integerRangeTest(in: T.min ..< (T.min + 10))
integerRangeTest(in: T.min ... (T.min + 10))
integerRangeTest(in: (T.max - 10) ..< T.max)
integerRangeTest(in: (T.max - 10) ... T.max)
if T.bitWidth == 8 {
  integerRangeTest(in: T.min ..< T.max)
  integerRangeTest(in: T.min ... T.max)
}
```

By the way, all the recent changes have been squashed into the "Consolidate `_stdlib_random` functions" commit.

benrimmington commented on stdlib/public/core/FloatingPoint.swift.gyb:
```diff
@@ -2374,6 +2374,111 @@ extension BinaryFloatingPoint {
+%   exampleRange = '10.0..<20.0' if Range == 'Range' else '10.0...20.0'
+extension BinaryFloatingPoint
+where Self.RawSignificand : FixedWidthInteger,
+      Self.RawSignificand.Stride : SignedInteger & FixedWidthInteger,
```
Why does the `RawSignificand.Stride` need to be `FixedWidthInteger`?

benrimmington commented on stdlib/public/core/FloatingPoint.swift.gyb:
```diff
@@ -2374,6 +2374,111 @@ extension BinaryFloatingPoint {
+  public static func random<T: RandomNumberGenerator>(
+    in range: ${Range}<Self>,
+    using generator: inout T
+  ) -> Self {
```
Here's another attempt at removing GYB code.

```swift
extension BinaryFloatingPoint where ... {

  public static func random<T: RandomNumberGenerator>(
    in range: ${Range}<Self>,
    using generator: inout T
  ) -> Self {
    _precondition(
      !range.isEmpty,
      "Can't get random value with an empty range"
    )
    let unitRandom = _unitRandom(using: &generator)
    let lowerBound = range.lowerBound
    let upperBound = range.contains(range.upperBound)
      ? range.upperBound
      : range.upperBound.nextDown
    return (upperBound - lowerBound) * unitRandom + lowerBound
  }
}
```

The internal API would be outside the `% for Range in ['Range', 'ClosedRange']:` block.

```swift
extension BinaryFloatingPoint where ... {

  // Returns a random value in the `0.0 ... 1.0` unit interval,
  // using raw values which can be represented exactly.
  internal static func _unitRandom<T: RandomNumberGenerator>(
    using generator: inout T
  ) -> Self {
    let max = (significandBitCount == RawSignificand.bitWidth - 1)
      ? RawSignificand.max
      : (1 as RawSignificand) << (significandBitCount + 1)
    let raw = RawSignificand.random(in: 0 ... max, using: &generator)
    return Self(exactly: raw)! / Self(exactly: max)!
  }
}
```

The [`significandBitCount`](https://developer.apple.com/documentation/swift/binaryfloatingpoint/1846714-significandbitcount) can be `Int.max` for extensible floating-point types. I'm not sure if we need to worry about that.

benrimmington replied:
I think `1.0 / ulpOfOne` gives the maximum precision needed.

```diff
-    let max = (significandBitCount == RawSignificand.bitWidth - 1)
-      ? RawSignificand.max
-      : (1 as RawSignificand) << (significandBitCount + 1)
+    let max = RawSignificand(1.0 / ulpOfOne)
```

Azoy replied:
```swift
let upperBound = range.contains(range.upperBound)
      ? range.upperBound
      : range.upperBound.nextDown
    return (upperBound - lowerBound) * unitRandom + lowerBound
```

I'm not sure that logic works. Given `0.0 ..< 2.0`, `upperBound = 1` so `(1 - 0) * unit + 0` will only produce values within `0.0 ... 1.0` whereas we need values from `0.0 ..< 2.0`.

benrimmington replied:
@Azoy, `range.contains(range.upperBound)` is testing for a closed range; and `range.upperBound.nextDown` converts a half-open range into a closed range.

Given `0.0 ..< 2.0`, the local `upperBound` and the delta will be `1.9999999999999998`; I've tested this and it appears to work as intended.

Azoy replied:
Oh my apologies, I assumed `nextDown` would return the next integer down.

benrimmington replied:
The non-branching version would be:

```swift
extension BinaryFloatingPoint where ... {

  public static func random<T: RandomNumberGenerator>(
    in range: ${Range}<Self>,
    using generator: inout T
  ) -> Self {
    _precondition(
      !range.isEmpty,
      "Can't get random value with an empty range"
    )
    let unitRandom = _unitRandom(using: &generator)
    let lowerBound = range.lowerBound
%   if 'Closed' in Range:
    let upperBound = range.upperBound
%   else:
    let upperBound = range.upperBound.nextDown
%   end
    return (upperBound - lowerBound) * unitRandom + lowerBound
  }
}
```

Azoy committed:
Clean up integerRangeTest
